aten/src/THC/generic/THCTensorMath.cu

```diff
@@ -116,30 +118,42 @@ void THCTensor_(catArray)(THCState *state, THCTensor *result,
 
   THArgCheck(numInputs > 0, 3, "invalid number of inputs %d", numInputs);
   THArgCheck(dimension >= 0, 4, "invalid dimension %d", dimension);
-  
-  std::vector<int64_t> size(nDims);
-  
+
   // Compute size of the result in the cat dimension
   int64_t cat_dim_size = 0;
   for (int i = 0; i < numInputs; i++) {
     THCTensor *tensor = inputs[i];
-    if (should_skip(tensor)) {
+    if (should_skip(tensor) && !pad) {
       continue;
     }
-    THCTensor_(check_shape_except_dim)(state, notSkippedTensor, tensor, dimension);
+    THCTensor_(check_shape_except_dim)(state, notSkippedTensor, tensor, dimension, pad);
     cat_dim_size += THCTensor_(size)(state, tensor, dimension);
   }
 
   // Compute the size of the result
+  std::vector<int64_t> size(nDims);
   for (int dim = 0; dim < nDims; dim++) {
     int64_t result_dim_size = THCTensor_(size)(state, notSkippedTensor, dim);
+    if (pad) {
+      for (int i = 0; i < numInputs; i++) {
+        THCTensor* tensor = inputs[i];
+        if (THCTensor_(size)(state, tensor, dim) > result_dim_size) {
+          result_dim_size = THCTensor_(size)(state, tensor, dim);
+        }
+      }
+    }
     if (dim == dimension) {
       result_dim_size = cat_dim_size;
     }
     size[dim] = result_dim_size;
   }
   THCTensor_(resize)(state, result, size, {});
 
+  // filled with pad values if required
+  if (pad) {
+    THCTensor_(fill)(state, result, pad_value);
```

Comment:
this is pretty expensive (particularly if you don't even need to pad because the sizes are all proper).  Not a big deal in the CPU, but this seems pretty wasteful in CUDA.  Can this be moved inside the efficient kernel?